### DISCUSSION TITLE
제네릭에 대해 설명해주세요.

### DISCUSSION THREAD

**Discussion by J-I-H-O** [2024-woowacourse-study/level-interview#83] [Java]
사용하는 이유까지

**Comment by gyungchan-jo** [comment]
자바 제네릭은 타입 안정성을 제공하고 재사용성을 높이기 위한 기능으로, 컬렉션 클래스나 메서드 등에서 다양한 타입을 사용할 수 있도록 도와줍니다.

사용하는 이유:
1. 타입 안정성 보장: 제네릭을 사용하면 컴파일 시점에 타입 안정성을 보장할 수 있습니다.
2. 코드 재사용성 증가: 제네릭을 사용하면 여러 종류의 객체에 대해 동일한 코드를 사용할 수 있습니다.
3. 유연성 제공: 제네릭을 사용하면 다양한 타입의 객체를 다룰 수 있으므로, 보다 유연한 코드를 작성할 수 있습니다.

**Comment by le2sky** [comment]
자바에서 제네릭이란 데이터의 타입을 일반화하는 것을 의미합니다.
클래스와 메서드 내부에서 사용할 데이터 타입을 컴파일 시에 미리 지정하는 방법입니다.
타입에 대한 정의를 외부로 위임하여, 타입에 대한 유연성과 안정성을 확보할 수 있습니다. 

과거에는 여러 타입을 사용하는 대부분의 클래스나 메서드에서 인수, 반환값으로 Object 타입을 사용했습니다.
하지만, 이 경우에는 Object 객체를 다시 원하는 타입으로 변환해야 했는데, 이 과정에서 오류가 발생할 가능성이 있었습니다.
JDK 1.5부터 도입된 제네릭을 사용하면, 컴파일 시에 미리 타입이 정해지므로, 타입 검사나 타입 변환과 같은 번거로운 작업을 생략할 수 있습니다.

또한, 이전 자바 버전의 코드와 호환성을 위해서 컴파일되면 제네릭 타입은 사라지게 됩니다. 이를 제네릭 타입 소거(Erasure)라고 합니다.
